pybricksdev/connections/pybricks.py

```diff
@@ -37,13 +45,259 @@
+        pass
+
+    @abc.abstractmethod
+    async def get_firmware_version(self) -> Version:
+        pass
+
+    @abc.abstractmethod
+    async def get_protocol_version(self) -> Version:
+        pass
+
+    @abc.abstractmethod
+    async def get_hub_type(self) -> Tuple[HubKind, int]:
+        pass
+
+    @abc.abstractmethod
+    async def get_hub_capabilities(self) -> Tuple:
+        pass
+
+    @abc.abstractmethod
+    async def send_command(self, command: bytes) -> None:
+        pass
+
+    @abc.abstractmethod
+    async def set_service_handler(self, callback: Callable) -> None:
+        pass
+
+
+class _BLETransport(_Transport):
+    _device: BLEDevice
+    _client: BleakClient
+
+    def __init__(self, device: BLEDevice):
+        self._device = device
+
+    async def connect(self, disconnected_callback: Callable) -> None:
+        def handle_disconnect(_: BleakClient):
+            disconnected_callback()
+
+        self._client = BleakClient(
+            self._device, disconnected_callback=handle_disconnect
+        )
+
+        logger.info(f"Connecting to {self._device.name}")
+        await self._client.connect()
+
+    async def disconnect(self) -> None:
+        await self._client.disconnect()
+
+    async def get_firmware_version(self) -> Version:
+        fw_version = await self._client.read_gatt_char(FW_REV_UUID)
+        return Version(fw_version.decode())
+
+    async def get_protocol_version(self) -> Version:
+        protocol_version = await self._client.read_gatt_char(SW_REV_UUID)
+        return semver.VersionInfo.parse(protocol_version.decode())
+
+    async def get_hub_type(self) -> Tuple[HubKind, int]:
+        pnp_id = await self._client.read_gatt_char(PNP_ID_UUID)
+        return unpack_pnp_id(pnp_id)[2:4]
+
+    async def get_hub_capabilities(self) -> Tuple:
+        caps = await self._client.read_gatt_char(PYBRICKS_HUB_CAPABILITIES_UUID)
+        return unpack_hub_capabilities(caps)
+
+    async def send_command(self, command: bytes) -> None:
+        await self._client.write_gatt_char(
+            PYBRICKS_COMMAND_EVENT_UUID, command, response=True
+        )
+
+    async def send_legacy(self, data: bytes) -> None:
+        await self._client.write_gatt_char(NUS_RX_UUID, data, response=False)
+
+    async def set_service_handler(self, callback: Callable) -> None:
+        def handler(_, data):
+            callback(data)
+
+        await self._client.start_notify(PYBRICKS_COMMAND_EVENT_UUID, handler)
+
+    # Only used for protocol version < "1.3.0"
+    async def set_nus_handler(self, callback: Callable) -> None:
+        def handler(_, data):
+            callback(data)
+
+        await self._client.start_notify(NUS_TX_UUID, handler)
+
+
+class _USBTransport(_Transport):
+    _device: USBDevice
+    _disconnected_callback: Callable
+    _ep_in: Endpoint
+    _ep_out: Endpoint
+    _notify_callbacks = {}
+    _monitor_task: asyncio.Task
+    _response_event = asyncio.Event()
+    _response: int
```

Comment:
An event + value sounds like we want an `asyncio.Future()` or `asyncio.Queue()` instead.